### PR TITLE
Watch xUnit processes for dumps in RunTests

### PR DESCRIPTION
**Pull request by jasonmalinowski**
We will launch ProcDump to monitor the existing xUnit processes. The dump files are written out to the same directory as results themselves.

*Note:* this currently has a commit that will intentionally break tests. That's to verify this actually works on the Jenkins machines end-to-end.

*Review:* @dotnet/roslyn-infrastructure, @dotnet/roslyn-ide, @rchande

### CODE REVIEW

File: src/Workspaces/CoreTest/AssemblyAttributes.cs [hunk 1]
```diff
@@ -1,4 +0,0 @@
-﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
Comment (Pilchie) [inline src/Workspaces/CoreTest/AssemblyAttributes.cs]: But this file - it was so useful!

**Comment by rchande** [comment]
:+1:

File: src/Workspaces/CoreTest/AssemblyAttributes.cs (hunk 1 above)
Comment (jasonmalinowski) [inline src/Workspaces/CoreTest/AssemblyAttributes.cs]: Not to the compiler it wasn't.

**Comment by Pilchie** [comment]
:+1:, assuming it works.

**Comment by jmarolf** [comment]
:+1:, assuming it works.

File: src/Tools/Source/RunTests/CrashDumps.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            var processStart = new ProcessStartInfo();
+
+            processStart.Arguments = GetProcDumpArgumentsForMonitoring(processToMonitor.Id, outputPath);
+            processStart.CreateNoWindow = true;
+            processStart.FileName = procDumpPath;
+            processStart.UseShellExecute = false;
+
+            var process = Process.Start(processStart);
+
+            return true;
```
Comment (brettfo) [inline src/Tools/Source/RunTests/CrashDumps.cs]: Necessary?  You don't appear to do anything with it.

File: src/Tools/Source/RunTests/CrashDumps.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
+﻿using System;
```
Comment (brettfo) [inline src/Tools/Source/RunTests/CrashDumps.cs]: Copyright.

File: src/Tools/Source/RunTests/CrashDumps.cs (hunk 2 above)
Comment (jasonmalinowski) [inline src/Tools/Source/RunTests/CrashDumps.cs]: I couldn't willingly write a Try* method that didn't return bool. It's immoral.

File: src/Tools/Source/RunTests/CrashDumps.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            // -ma: create a full memory dump
+            //
+            // without -g, procdump will not catch unhandled managed exception since CLR will always
+            // handle unhandled exception. for procdump point of view, there is no such thing as unhandled
+            // exception for managed app.
+
+            return $"-accepteula -g -e -b -ma {pid} {outputPath}";
+        }
+
+        private const string DotNetContinuousIntegrationProcDumpLocation = @"C:\Sysinternals\Procdump.exe";
```
Comment (brettfo) [inline src/Tools/Source/RunTests/CrashDumps.cs]: Are the Jenkins machines guaranteed to have this?  Maybe if this file isn't found but we've detected (somehow?) that we're a Jenkins run, fail miserably?

File: src/EditorFeatures/CSharpTest/CrashHardTest.cs [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.UnitTests
+{
+    public class CrashHardTest
+    {
+        [Fact, Trait(Traits.Feature, "Crash")]
+        public void Crash()
+        {
+            ExceptionUtilities.FailFast(new Exception("Goodbye!"));
```
Comment (brettfo) [inline src/EditorFeatures/CSharpTest/CrashHardTest.cs]: Would be nice to also verify that `Debug.Assert()` causes a dump.

Comment (jaredpar) [inline src/EditorFeatures/CSharpTest/CrashHardTest.cs]: Shouldn't this really be saying `"Hello"` instead?  After all this is going to pop up in a developer's email eventually, may as well have a greeting to go with it vs. a dismissal.  :trollface:

File: src/Tools/Source/RunTests/CrashDumps.cs (hunk 4 above)
Comment (jaredpar) [inline src/Tools/Source/RunTests/CrashDumps.cs]: Why not use `\\live.sysinternals.com\tools\procdump.exe`

Comment (jaredpar) [inline src/Tools/Source/RunTests/CrashDumps.cs]: There may be some lag on that network share.  So maybe copy it from that share to `Binaries\Debug` and run it from there afterwards.

File: src/EditorFeatures/CSharpTest/CrashHardTest.cs (hunk 5 above)
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest/CrashHardTest.cs]: Well, if this works, by the power of Git:

you_didnt_see_anything

File: src/Tools/Source/RunTests/CrashDumps.cs (hunk 4 above)
Comment (jasonmalinowski) [inline src/Tools/Source/RunTests/CrashDumps.cs]: @tannergooding assures us this file exists. We can expand this with fancier stuff eventually.

Comment (jaredpar) [inline src/Tools/Source/RunTests/CrashDumps.cs]: Even if it's on Jenkins I would like to have the dumps locally if things crash.  Can deal with this later though.

Comment (jasonmalinowski) [inline src/Tools/Source/RunTests/CrashDumps.cs]: Well, we really should put it in a package or something. \\live.sysinternals.com assumes you have a network. But I don't want to block on this.

**Comment by balajikris** [comment]
:+1:

I thought we had this for a long time. not sure where it went, good to have it back :smile:

**Comment by jasonmalinowski** [comment]
@dotnet-bot, retest prtest/win/dbg/unit32 please. The test failed as expected, but artifacts already disappeared so I can't verify it worked.

**Comment by jasonmalinowski** [comment]
I've verified the intentionally-failed build did get a dump. I've removed that commit now for actual submission.

**Pull request merged**